src/main/java/apoc/uuid/UuidHandler.java

```diff
@@ -0,0 +1,167 @@
+
+    public static final String APOC_UUID = "apoc.uuid";
+    static ConcurrentHashMap<String, Map<String, Object>> uuid = new ConcurrentHashMap(map("", map()));
+    private static GraphProperties properties;
+    private final Log log;
+
+    public static final String NOT_ENABLED_ERROR = "UUID have not been enabled." +
+            " Set 'apoc.uuid.enabled=true' in your neo4j.conf file located in the $NEO4J_HOME/conf/ directory.";
+
+    public UuidHandler(GraphDatabaseAPI api, Log log) {
+        properties = api.getDependencyResolver().resolveDependency(EmbeddedProxySPI.class).newGraphPropertiesProxy();
+        this.log = log;
+    }
+
+    public static void checkEnabled() {
+        if (properties == null) {
+            throw new RuntimeException(NOT_ENABLED_ERROR);
+        }
+    }
+
+    public static Map<String, Object> add(String name, List<String> labels, Map<String, Object> params) {
+        checkEnabled();
+        checkConstraintUuid(labels);
+        return updateUuid(name, map("labels", labels, "params", params, "paused", false));
+    }
+
+    private synchronized static Map<String, Object> updateUuid(String name, Map<String, Object> value) {
+        checkEnabled();
+        try (Transaction tx = properties.getGraphDatabase().beginTx()) {
+            uuid.clear();
+            String uuidProperty = (String) properties.getProperty(APOC_UUID, "{}");
+            uuid.putAll(Util.fromJson(uuidProperty, Map.class));
+            Map<String, Object> previous = null;
+            if (name != null) {
+                previous = (value == null) ? uuid.remove(name) : uuid.put(name, value);
+                if (value != null || previous != null) {
+                    properties.setProperty(APOC_UUID, Util.toJson(uuid));
+                }
+            }
+            tx.success();
+            return previous;
+        }
+    }
+
+    @Override
+    public Object beforeCommit(TransactionData data) throws Exception {
+        executeUuid(data, "before");
+        return null;
+    }
+
+    @Override
+    public void afterCommit(TransactionData data, Object state) {
+        executeUuid(data, "after");
+    }
+
+    @Override
+    public void afterRollback(TransactionData data, Object state) {
+        executeUuid(data, "roolback");
+    }
+
+    private void executeUuid(TransactionData txData, String phase) {
+        if (uuid.containsKey("")) updateUuid(null,null);
+        GraphDatabaseService db = properties.getGraphDatabase();
+        Map<String, String> exceptions = new LinkedHashMap<>();
+
+        uuid.forEach((key, data) -> {
+            if (data.get("paused").equals(false)) {
+                try (Transaction tx = db.beginTx()) {
+                    List<String> labels = (List<String>) data.get("labels");
+                    labels.forEach(label -> {
+                        ResourceIterator<Object> result = db.execute(String.format("MATCH (n:%s) WHERE not exists(n.uuid) RETURN n", label)).columnAs("n");
```

Comment:
perché facciamo questa query?
non possiamo fare `tx.createdNodes()` (per accedere a tutti i nodi creat) e `tx.assignedNodeProperties()` (per verificare che accidentalmente non sia stata cancellata la proprieta e in tal caso ripristinarla se era presente) nella beforeCommit e verificare se per questi è necessario aggiungere l'uuid?